### PR TITLE
Add cppyy

### PR DESCRIPTION
**Pull request by saraedum**
* [x] Use sources from PyPA instead of the bitbucket snapshots.
* [x] Add some non-import test for the final cppyy package
* [x] Build for all Python flavours again

### CODE REVIEW

**Comment by saraedum** [comment]
@isuruf: I pushed some mostly boilerplate recipes here to get the process started.

**Comment by saraedum** [comment]
@isuruf: I added you as a maintainer. I guess that's fine since you were helping quite a bit with this?

**Comment by isuruf** [comment]
Sure

**Comment by saraedum** [comment]
(we first have to make it build of course…)

**Comment by saraedum** [comment]
@isuruf: Linux failed in the same way again. OSX "passed" but there were actually errors…no idea what's going on there.

File: recipes/cppyy-cling/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+    - cmake
+  host:
+    - python
+    - pip
+    - zlib
+    - clangdev={{ clang_version }}
```
Comment (chrisburr) [inline recipes/cppyy-cling/meta.yaml]: Having a space here is important for conda-forge to work correctly:
```suggestion
    - clangdev ={{ clang_version }}
```

File: recipes/cppyy-cling/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    - clang_variant * cling_{{ cling_patches_version }}
+    - llvmdev==5
```
Comment (chrisburr) [inline recipes/cppyy-cling/meta.yaml]: Same
```suggestion
    - llvmdev ==5
```

File: recipes/cppyy-cling/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    # Copied over from conda-forge's cling recipe:
+    # FIXME: The generated allDict.cxx.pch is dependent on version of the C++
+    # headers used As it is shipped in the macOS binaries we have to use the
+    # same one as in the compilers package
+    - libcxx =4  # [osx]
+  run:
+    - compilers
+    - python
+    # Copied over from conda-forge's cling recipe:
+    - libcxx=4  # [osx]
```
Comment (chrisburr) [inline recipes/cppyy-cling/meta.yaml]:
```suggestion
    - libcxx =4  # [osx]
```

**Comment by saraedum** [comment]
Thanks for the comments @chrisburr. I copied these over verbatim from https://github.com/conda-forge/root-feedstock/blob/master/recipe/meta.yaml.

**Comment by saraedum** [comment]
@isuruf: With the setup from root-feedstock I don't get the features.h error anymore. If I understood you correctly, there is no way to build cppyy-cling without a `python` dependency? It's annoying that I have to build it three times now (though it's much faster to build without clang & llvm.)

**Comment by isuruf** [comment]
> If I understood you correctly, there is no way to build cppyy-cling without a python dependency?

Yeah, that's not supported by conda-build yet.

**Comment by saraedum** [comment]
> > If I understood you correctly, there is no way to build cppyy-cling without a python dependency?
> 
> Yeah, that's not supported by conda-build yet.

Do you have a reference for this so I can put this as a comment in the recipe?

**Comment by saraedum** [comment]
@conda-forge-linter: I do have a test there…

File: recipes/cpycppyy/meta.yaml [hunk 4]
```diff
@@ -29,7 +29,9 @@ requirements:
 
 test:
   commands:
-    - ls {{ SP_DIR }}/libcppyy.cpython*.so
+    # This does not work as SP_DIR is the directory at build time.
+    # - ls {{ SP_DIR }}/libcppyy.cpython*.so
```
Comment (isuruf) [inline recipes/cpycppyy/meta.yaml]: use `$SP_DIR` instead of the jinja variable

Comment (saraedum) [reply recipes/cpycppyy/meta.yaml]: I see. I had understood the documentation that these variables are only available at build time.

File: recipes/cppyy-cling/build.sh [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+declare -a CMAKE_PLATFORM_FLAGS
+if [ "$(uname)" == "Linux" ]; then
+    CMAKE_PLATFORM_FLAGS+=("-DCMAKE_AR=${GCC_AR}")
+    CMAKE_PLATFORM_FLAGS+=("-DCLANG_DEFAULT_LINKER=${LD_GOLD}")
+    CMAKE_PLATFORM_FLAGS+=("-DDEFAULT_SYSROOT=${PREFIX}/${HOST}/sysroot")
+    CMAKE_PLATFORM_FLAGS+=("-DRT_LIBRARY=${PREFIX}/${HOST}/sysroot/usr/lib/librt.so")
+
+    # Fix up CMake for using conda's sysroot
+    # See https://docs.conda.io/projects/conda-build/en/latest/resources/compiler-tools.html?highlight=cmake#an-aside-on-cmake-and-sysroots
+    CMAKE_PLATFORM_FLAGS+=("-DCMAKE_TOOLCHAIN_FILE=${RECIPE_DIR}/cross-linux.cmake")
```
Comment (isuruf) [inline recipes/cppyy-cling/build.sh]: This is unnecessary

Comment (saraedum) [reply recipes/cppyy-cling/build.sh]: I copied this over from https://github.com/conda-forge/root-feedstock/blob/master/recipe/build.sh#L22. So it should be removed there as well? @chrisburr

**Comment by saraedum** [comment]
Finally this seems to work (still waiting for the Linux pipeline to finish). There is a warning when starting cppyy:
```
Error in <UnknownClass::InitInterpreter()>: LLVM SYMBOLS ARE EXPOSED TO CLING! This will cause problems; please hide them or dlopen() them after the call to TROOT::InitInterpreter()!
```

This as from Cling. Not sure whether this is really a problem but I'd probably like to improve this in the feedstock. Also this certainly does not pass all the tests that come with cppyy as patches to clang & llvm are missing.

**Comment by saraedum** [comment]
@isuruf would you mind to review this once more?

File: recipes/cppyy-cling/build.sh [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+declare -a CMAKE_PLATFORM_FLAGS
+if [ "$(uname)" == "Linux" ]; then
+    CMAKE_PLATFORM_FLAGS+=("-DCMAKE_AR=${GCC_AR}")
+    CMAKE_PLATFORM_FLAGS+=("-DCLANG_DEFAULT_LINKER=${LD_GOLD}")
+    CMAKE_PLATFORM_FLAGS+=("-DDEFAULT_SYSROOT=${PREFIX}/${HOST}/sysroot")
+    CMAKE_PLATFORM_FLAGS+=("-DRT_LIBRARY=${PREFIX}/${HOST}/sysroot/usr/lib/librt.so")
+
+    # Hide symbols from LLVM/clang to avoid conflicts with other libraries
+    for lib_name in $(ls $PREFIX/lib | grep -E 'lib(LLVM|clang).*\.a'); do
+        export CXXFLAGS="${CXXFLAGS} -Wl,--exclude-libs,${lib_name}"
```
Comment (isuruf) [inline recipes/cppyy-cling/build.sh]: Shouldn't this be LDFLAGS instead of CXXFLAGS? Changing this would probably fix your warning

Comment (saraedum) [reply recipes/cppyy-cling/build.sh]: This is from root's build script. Let's try.

Comment (saraedum) [reply recipes/cppyy-cling/build.sh]: Then again, this won't help on macOS.

Comment (saraedum) [reply recipes/cppyy-cling/build.sh]: Actually, the warning now only shows on macOS, so I guess there's no need to change this.

**Comment by saraedum** [comment]
@isuruf @chrisburr: Imho this can be merged now :)

**Pull request merged**